google/cloud/spanner/numeric.h

```diff
@@ -0,0 +1,269 @@
+ * 29 decimal digits of integer precision (kIntPrec) and 9 decimal digits of
+ * fractional precision (kFracPrec).
+ *
+ * So, the range of a `Numeric` is -99999999999999999999999999999.999999999
+ * to 99999999999999999999999999999.999999999.
+ *
+ * A `Numeric` can be constructed from, and converted to a `std::string`, a
+ * `double`, or any integral type. See the `MakeNumeric()` factory functions,
+ * the `ToString()` member function, and the `ToDouble()`/`ToInteger<T>()`
+ * free functions.
+ *
+ * `Numeric` values can be copied/assigned/moved, compared for equality, and
+ * streamed.
+ */
+class Numeric {
+ public:
+  /// Decimal integer and fractional precision of a `Numeric` value.
+  ///@{
+  static constexpr std::size_t kIntPrec = 29;
+  static constexpr std::size_t kFracPrec = 9;
+  ///@}
+
+  /// A zero value.
+  Numeric();
+
+  /// @name Regular value type, supporting copy, assign, move.
+  ///@{
+  Numeric(Numeric&&) = default;
+  Numeric& operator=(Numeric&&) = default;
+  Numeric(Numeric const&) = default;
+  Numeric& operator=(Numeric const&) = default;
+  ///@}
+
+  /**
+   * Conversion to a decimal-string representation of the `Numeric` in one
+   * of the following forms:
+   *   0                                      // value == 0
+   *   -?0.[0-9]{0,8}[1-9]                    // 0 < abs(value) < 1
+   *   -?[1-9][0-9]{0,28}(.[0-9]{0,8}[1-9])?  // abs(value) >= 1
+   *
+   * Note: The string never includes an exponent field.
+   */
+  ///@{
+  std::string const& ToString() const& { return rep_; }
+  std::string&& ToString() && { return std::move(rep_); }
+  ///@}
+
+  /// @name Relational operators
+  ///@{
+  friend bool operator==(Numeric const& a, Numeric const& b) {
+    return a.rep_ == b.rep_;
+  }
+  friend bool operator!=(Numeric const& a, Numeric const& b) {
+    return !(a == b);
+  }
+  ///@}
+
+  /// Outputs string representation of the `Numeric` to the provided stream.
+  friend std::ostream& operator<<(std::ostream& os, Numeric const& n) {
+    return os << n.ToString();
+  }
+
+ private:
+  friend StatusOr<Numeric> internal::MakeNumeric(std::string s);
+  explicit Numeric(std::string rep) : rep_(std::move(rep)) {}
+  std::string rep_;  // a valid and canonical NUMERIC representation
+};
+
+namespace internal {
+
+// Like `std::to_string`, but also supports Abseil 128-bit integers.
+template <typename T>
+std::string ToString(T value) {
+  return std::to_string(value);
+}
+std::string ToString(absl::int128 value);
+std::string ToString(absl::uint128 value);
+
+// Forward declarations.
+Status DataLoss(std::string message);
+StatusOr<Numeric> MakeNumeric(std::string s, int exponent);
+
+}  // namespace internal
+
+/**
+ * Construction from a string, in decimal fixed- or floating-point formats
+ * (where there must be digits either before or after the decimal point):
+ *
+ *    [-+]?[0-9]+(.[0-9]*)?([eE][-+]?[0-9]+)?
+ *    [-+]?.[0-9]+([eE][-+]?[0-9]+)?
```

Comment:
It may be helpful for human readers to see a few examples of of numeric strings so they don't have to interpret this more formal description.